beacon-chain/core/helpers/weak_subjectivity.go

```diff
@@ -95,3 +97,47 @@ func ComputeWeakSubjectivityPeriod(st iface.ReadOnlyBeaconState) (types.Epoch, e
 
 	return types.Epoch(wsp), nil
 }
+
+// IsWithinWeakSubjectivityPeriod verifies if a given weak subjectivity checkpoint is not stale i.e.
+// the current node is so far beyond, that a given state and checkpoint are not for the latest weak
+// subjectivity point. Provided checkpoint still can be used to double-check that node's block root
+// at a given epoch matches that of the checkpoint.
+//
+// Reference implementation:
+// https://github.com/ethereum/eth2.0-specs/blob/master/specs/phase0/weak-subjectivity.md#checking-for-stale-weak-subjectivity-checkpoint
+
+// def is_within_weak_subjectivity_period(store: Store, ws_state: BeaconState, ws_checkpoint: Checkpoint) -> bool:
+//    # Clients may choose to validate the input state against the input Weak Subjectivity Checkpoint
+//    assert ws_state.latest_block_header.state_root == ws_checkpoint.root
+//    assert compute_epoch_at_slot(ws_state.slot) == ws_checkpoint.epoch
+//
+//    ws_period = compute_weak_subjectivity_period(ws_state)
+//    ws_state_epoch = compute_epoch_at_slot(ws_state.slot)
+//    current_epoch = compute_epoch_at_slot(get_current_slot(store))
+//    return current_epoch <= ws_state_epoch + ws_period
+func IsWithinWeakSubjectivityPeriod(
+	currentEpoch types.Epoch, wsState iface.ReadOnlyBeaconState, wsCheckpoint *eth.WeakSubjectivityCheckpoint) (bool, error) {
+	// Make sure that incoming objects are not nil.
+	if wsState == nil || wsState.LatestBlockHeader() == nil || wsCheckpoint == nil {
+		return false, errors.New("invalid weak subjectivity state or checkpoint")
+	}
+
+	// Assert that state and checkpoint have the same root and epoch.
+	if !bytes.Equal(wsState.LatestBlockHeader().StateRoot, wsCheckpoint.StateRoot) {
```

Comment:
So here is the thing every state root is zeroed out after a block is processed. This check will never work for an actual beacon state since it will always be a zero-hash.

Reply:
Please see https://discord.com/channels/476244492043812875/730447046930071672/828595210056892446 I believe that there's an issue with this check, it should be comparing `wsState.LatestBlockHeader().HashTreeRoot()` not a state (as it is written in specs).

Reply:
Gotcha thanks for clarifying, although even with that the root derived from the block header will be inaccurate. Because it still has a zeroed out state root in the cached header. The specs should probably clarify this point properly, seems ambiguous here.

Reply:
Well, this check will be on embedded/incoming state -- so I guess it will have block root not zeroed i.e. initial state will be provided, it will have the root, and then before trying to sync from that root, system will check if the checkpoint (and corresponding incoming state) is stale -- and issue a warning that state is stale (root and epoch of checkpoint will still be validated against what is already known by the node).

Reply:
If it was the case, then it would be tricky to use it given that our state and slot transitions assume a zeroed out state root. Having a non-zero state root for the latest block header is usually a feature of an intermediate state so it might make syncing trickier, I guess will see what the specs clarify on here.

Reply:
What about `wsInitialState.LatestBeaconHeader().HashTreeRoot()`? I suppose it is not expected to be zeroed (and if it is not -- then I'll just need to file PR to spec pointing out that we are dealing with block root of the latest known block, not the state)?

Reply:
Yeap correct, the root is not expected to be zeroed out but it will be incorrect for the reasons we mentioned above. 
> then I'll just need to file PR to spec pointing out that we are dealing with block root of the latest known block, not the state

That makes sense to me, I took a quick look through the doc and they are definitely referring to block root in quite a few places, so I assume its a mistake in the spec.